### DISCUSSION TITLE
July release - SSO, conditional UI, and more (Beta)

### DISCUSSION THREAD

**Discussion by joebudi** [Budibase/budibase#2165] [Announcements]
The July release delivers on several popular requests from our community and contains four major updates, including;

1. SSO - OIDC support
2. Conditional UI
3. Themeable apps
4. Map SQL relationships within the builder

---

# SSO - OIDC support (free)
<img width="433" alt="sso dark" src="https://user-images.githubusercontent.com/49767913/127457768-a235b4f8-bb3f-4962-9209-af7d2cc1dc32.png">
OIDC is an open standard and decentralized authentication protocol. As a user, you will have infinite possibilities for SSO, including:

- Active Directory! - popular request
- Auth0
- Azure
- Okta
- OneLogin
- It’s also possible to add your own authentication! 

I know many of you will understand the benefits of SSO, but here’s a quick recap:

- No more password management
- Auto-provisioning for users
- Minimize password security risks
- Quicker sign up/sign-in process for users
- Use your SSO provider to control who can access Budibase

The initial work for this feature was completed by community member [Bernhard Hayden](https://github.com/burnoutberni). We’d like to publicly thank Bernhard for taking the time and effort to contribute to Budibase. The remaining work was completed by two new Budibase recruits; [Peter](https://github.com/PClmnt) and [Rory](https://github.com/Rory-Powell).

---

# Conditional UI
![conditions](https://user-images.githubusercontent.com/49767913/127457749-df16715d-852e-4f60-aa69-994158bc4436.png)
When building apps with Budibase, you will now have the ability to change your UI depending on the application's state. This new feature allows you to hide content from certain users, change the color of text depending on its value, and more.

Quick demo:
https://user-images.githubusercontent.com/49767913/127480993-beb592cb-d961-46e9-bd84-6ca39b700e14.mp4

---

# Dark mode apps 
![app themes](https://user-images.githubusercontent.com/49767913/127457829-e17d0e73-f9be-44a9-a69d-c0d2a1f428f5.png)
We strive to make it as easy as possible for you to build beautiful, user-friendly internal tools with Budibase. After this release, switch between 4 themes for your apps; lightest, light, dark, darkest, with just 2 clicks.

---

# SQL relationships
![sql relationships](https://user-images.githubusercontent.com/49767913/127457796-411695ec-95fa-4892-be72-977abb68478f.png)
Import tables from your SQL databases (with just 1 click), then map relationships between your tables within Budibase. This allows you to utilize the power of relationships within your Budibase applications. 

---

## Additional updates
- Unauthenticated SMTP
- Budibase drawer UI improvements
- [Layout update](https://github.com/Budibase/budibase/pull/1969)
- Lucene search improvements

---

# Coming up in the August release
- Dynamic forms and a Wizard component
- A better developer experience (JavaScript, transformers, and auto-complete)

---

**Watch release video**
[![Watch the release video](https://img.youtube.com/vi/LJHA2K8loaw/default.jpg)](https://www.youtube.com/watch?v=LJHA2K8loaw)

**Comment by oswaldB** [comment]
Hello,

How to upgrage to the new release ?

When I click update in the app, it says already up-to-date and I cannot see the conditional ui.
![image](https://user-images.githubusercontent.com/3023526/127477766-f0781bfb-d5c1-4163-b8f4-e50949b64047.png)


thank you

**Reply by shogunpurple** [reply]
@oswaldB you may have to upgrade your client library in your apps or create a new app.

See this comment: https://github.com/Budibase/budibase/discussions/2160#discussioncomment-1084598

**Reply by joebudi** [reply]
**Within** your app, you will notice a small refresh/update icon. Click that (it's the blue icon in the image below)
![Screenshot 2021-07-29 at 11 43 11](https://user-images.githubusercontent.com/49767913/127478528-237b2130-302b-4caf-8dfd-d3647b6d7aef.png)

**Reply by shogunpurple** [reply]
There’s also a possibility you are having some issues with your update server. Could you post the results of a

docker ps

on your server?

**Reply by oswaldB** [reply]
the cirlce arrow is working perfectly fine :)

**Reply by oswaldB** [reply]
It works perfectly @joebudi :) Thank you

**Comment by KernOil** [comment]
@shogunpurple With the new custom configurations, how do I go about hiding and unhiding components by clicking on a button?

I figure I will need two buttons one for hiding one for unhideing. what is the trigger point I can call when the button is pressed?

**Reply by kevmodrome** [reply]
Use some kind of boolean on your data. For example, have your "Show" button hidden on `data.paid === true` and your "Hide" button hidden on `data.paid === false`. Then set up an action on these buttons to change the value to true or false

**Reply by kevmodrome** [reply]
More detailed documentation on Conditional UI is coming tomorrow. I will make sure to ping you when it is available.

**Reply by KernOil** [reply]
That would be great thank you. this will bring so much more functionality to my app really looking forward to it!

**Reply by kevmodrome** [reply]
In the meantime, here's the setup from the release video:

<img width="1273" alt="CleanShot 2021-07-29 at 20 54 34@2x" src="https://user-images.githubusercontent.com/534488/127549916-5733c403-a6ce-441f-98f3-d5b70c33b03c.png">

**Reply by KernOil** [reply]
Thank you

**Reply by KernOil** [reply]
to me, this looks like it is a one-way click.  once the Boolean value changes can it be changed back?

**Reply by shogunpurple** [reply]
Hey @AndrewJRamay - they are not one way, they will update as data changes 👍

**Reply by kevmodrome** [reply]
@AndrewJRamay The conditions docs can be found here: https://docs.budibase.com/design/conditions if you have any questions please post them as separate questions in the the Help category: https://github.com/Budibase/budibase/discussions/categories/help

**Comment by Fabian1410** [comment]
Dark mode doesnt work, with updated apps :( 

Only with new ones :D

**Reply by joebudi** [reply]
Try clicking the blue refresh icon beside the preview icon.

**Reply by Fabian1410** [reply]
It say, already Upgraded …

**Comment by KernOil** [comment]
is there a trick to get the clear form action to work? in my button I have the data validated then I save the row and I would like the form cleared but nothing happens after the row is saved.

**Reply by KernOil** [reply]
is there a way I can just refresh the page? this would clear the form and display the new data in the table at the same time.
i have tried the custom CSS  in automation to do this but I cant get it to work.

**Reply by shogunpurple** [reply]
@AndrewJRamay have you set up the clear form action after your save? Could I see the way your actions are set up?

**Reply by KernOil** [reply]
Hi @shogunpurple the clear form is now working after the restart last night. however, that action is not refreshing the data table at the bottom. I would like the user to be able to see the new row they just created. the only way I can get this to show the new data is to refresh the page. 
![image](https://user-images.githubusercontent.com/80058238/127690254-e95a92c0-cd1f-4799-ab2c-fdc96c0a8efe.png)

**Comment by Manishkparmar80** [comment]
Can you pls share document on configuration of SSO on Active Directory.

**Reply by shogunpurple** [reply]
https://docs.budibase.com/admin/authentication

**Reply by Rory-Powell** [reply]
@Manishkparmar80 do you mean azure active directory or microsoft server active directory?

for azure ad:
as martin linked about we have https://docs.budibase.com/admin/authentication/azure

for on premise active directory:
you can use federation services to connect using openid. 
we don't have a guide for this yet but there is an external guide here https://wiki.resolution.de/doc/openid-oauth-authentication/latest/setup-guides/adfs-setup-guide

**Reply by Manishkparmar80** [reply]
@Rory-Powell  I mean On Premise Active Directory.  Presently we do not have ADFS so we will setup new ADFS and try it out.

Presently in our environment many application & appliance are integrated with AD using LDAP for SSO functionality.    After successful LDAP synchronization with AD, Entire AD Users / organization unit gets sync with Application and application can use this Sync data as User Master (username, dept, email id, designation, emp id, DOJ, area, location etc.) and use it for many application functionality like role management, workflow and more .   Once new user get added into AD then Application user master automatically get updated with this newly added user using auto sync feature of LDAP. or if any user information gets modified into AD then Application user master automatically get updated with modified detail of user using auto sync feature of LDAP.

How to achieve this by newly added SSO feature of OIDC.

**Reply by KernOil** [reply]
@Rory-Powell for AD on-premise this  
"for on premise active directory:
you can use federation services to connect using openid.
we don't have a guide for this yet but there is an external guide here https://wiki.resolution.de/doc/openid-oauth-authentication/latest/setup-guides/adfs-setup-guide"

is my best guide for setup am I reading that correctly?

**Reply by Rory-Powell** [reply]
@Manishkparmar80 thanks for clarifying. The SSO feature in Budibase doesn't provide synchronization, right now it is just for login. In future we will add support for user sync (provisioning, updating, deprovisioning etc) using LDAP or SCIM, which will be a feature that will sit alongside SSO.

**Reply by Rory-Powell** [reply]
@AndrewJRamay correct that's the best guide to follow, it covers configuration steps that will be similar to Budibase. I recommend having a read through our own oidc docs before proceeding so you are familiar with the required attributes https://docs.budibase.com/admin/authentication/oidc

**Reply by Manishkparmar80** [reply]
@Rory-Powell @joebudi  if SSO function is only for managing login then how other functionality like user role to app , workflow, approvals , email notification etc would be managed without user master.

**Reply by mjashanks** [reply]
Hey @Manishkparmar80. We will implement LDAP sync, but not for a few months - SSO (authentication) is the first step towards this.

When we do implement it, it will come under a paid plan. Closer to the time, we'd love to engage with you on your needs around this to help us understand the uses.

**Comment by sergey-dme** [comment]
Really enjoying new functionality. Conditional UI is very powerful. It also allows to not duplicate same "base" screen changing the Data Provider filter, form fields and buttons as I did before, I think dynamic UI is most wanted feature for many users. Just an Idea - it would be great to see Background color option in Update Setting section (use-case: creating like a color "labels" using different filed conditions)

Love BB!

**Comment by bloomt** [comment]
Can someone help me update to this version. My version is showing "Current Version: 0.9.100" with no updated. I'm guessing im on the mast er branch but im interested in trying out some of these features.

I have run "budi hosting --update" but it did not change the version so I am guessing that I need a flag or config change to use the beta branch.

**Reply by shogunpurple** [reply]
@bloomt you are on the latest version. You should now have all the latest features.

**Comment by andyburgessmd** [comment]
I'm trying to get the OIDC SSO setup with OneLogin and have configured it to the instructions given. However when trying to authenticate as a user it end with a 403 error from the budibase.app (cloud) endpoint and the error in the url is `error=invalid_scope&error_description=some%20of%20requested%20scopes%20are%20not%20whitelisted`

Anyone had this and been able to fix?
